File: packages/web/src/game/physics/FlightModel.ts
// ---------------------------------------------------------------------------
// Flight-model physics — runs once per frame
//
// All units are SI internally:
//   position: metres, velocity: m/s, angles: radians, forces: newtons
// ---------------------------------------------------------------------------

import * as THREE from "three";
import { density } from "./Atmosphere";
import {
  GRAVITY,
  SEA_LEVEL_DENSITY,
  CL_ALPHA_SLOPE,
  STALL_ALPHA,
  CL_MAX,
  SIDE_FORCE_COEFF,
  AERO_YAW_COEFF,
  Q_CRUISE,
  PROP_PLANE,
} from "@/lib/constants";

// ── helpers ────────────────────────────────────────────────────────────────

/** Clamp a value between min and max. */
function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, v));
}

/**
 * Simplified lift-coefficient model.
 * Linear up to the stall angle, then drops off.
 */
export function liftCoefficient(alpha: number): number {
  const absAlpha = Math.abs(alpha);
  if (absAlpha <= STALL_ALPHA) {
    return CL_ALPHA_SLOPE * alpha;
  }
  // Post-stall: Cl drops linearly back toward zero
  const sign = Math.sign(alpha);
  const postStallFraction = 1 - (absAlpha - STALL_ALPHA) / (Math.PI / 2 - STALL_ALPHA);
  return sign * CL_MAX * clamp(postStallFraction, 0, 1);
}

// ── state type ─────────────────────────────────────────────────────────────

export interface AircraftState {
  position: THREE.Vector3;
  /** Euler rotation (pitch, yaw, roll) in intrinsic order "YXZ" */
  rotation: THREE.Euler;
  velocity: THREE.Vector3;
  airspeed: number;
  altitude: number;
  heading: number;
  throttle: number;
  angularVelocity: THREE.Vector3; // (pitchRate, yawRate, rollRate)
}

// ── control input ──────────────────────────────────────────────────────────

export interface ControlInput {
  pitch: number; // -1 (nose down) to +1 (nose up)
  roll: number; // -1 (left) to +1 (right)
  yaw: number; // -1 (left) to +1 (right)
  throttle: number; // 0 – 1 (absolute)
}

// ── integrator ─────────────────────────────────────────────────────────────

/**
 * Advance the flight model by `dt` seconds.
 *
 * Mutates `state` in place and returns it for convenience.
 */
export function stepFlightModel(
  state: AircraftState,
  input: ControlInput,
  dt: number,
): AircraftState {
  const {
    wingArea: S,
    mass,
    maxThrust,
    cd0: Cd0,
    oswaldEfficiency: e,
    aspectRatio: AR,
    pitchRate: maxPitch,
    rollRate: maxRoll,
    yawRate: maxYaw,
  } = PROP_PLANE;

  // Clamp dt to avoid physics explosions after tab-away
  const safeDt = Math.min(dt, 0.05);

  // ── 1. Rotation (control surfaces) ────────────────────────────────────

  // Build a quaternion from the current Euler so we can rotate vectors
  const quat = new THREE.Quaternion().setFromEuler(state.rotation);

  // Local axes in world space
  const forward = new THREE.Vector3(0, 0, -1).applyQuaternion(quat);
  const up = new THREE.Vector3(0, 1, 0).applyQuaternion(quat);
  const right = new THREE.Vector3(1, 0, 0).applyQuaternion(quat);

  // Apply angular velocities (exponential map — small angle approx is fine
  // for the per-frame deltas we deal with)
  const pitchDelta = input.pitch * maxPitch * safeDt;
  const yawDelta = input.yaw * maxYaw * safeDt;
  const rollDelta = input.roll * maxRoll * safeDt;

  const dq = new THREE.Quaternion();
  // Pitch around local right axis
  dq.setFromAxisAngle(right, pitchDelta);
  quat.premultiply(dq);
  // Yaw around local up axis
  dq.setFromAxisAngle(up, -yawDelta);
  quat.premultiply(dq);
  // Roll around local forward axis
  dq.setFromAxisAngle(forward, -rollDelta);
  quat.premultiply(dq);

  quat.normalize();
  state.rotation.setFromQuaternion(quat, "YXZ");

  // Recompute all local axes after rotation update
  forward.set(0, 0, -1).applyQuaternion(quat);
  up.set(0, 1, 0).applyQuaternion(quat);
  right.set(1, 0, 0).applyQuaternion(quat);

  // ── 2. Aerodynamic forces ─────────────────────────────────────────────

  const altitude = Math.max(state.position.y, 0);
  const rho = density(altitude);

  // Airspeed = magnitude of velocity
  const V = state.velocity.length();
  const dynamicPressure = 0.5 * rho * V * V; // q

  // Angle of attack: angle between forward vector and velocity
  let alpha = 0;
  if (V > 1) {
    const velDir = state.velocity.clone().normalize();
    // AoA is the angle in the plane of symmetry: use dot with forward and up
    const dotFwd = velDir.dot(forward);
    const dotUp = velDir.dot(up);
    alpha = Math.atan2(-dotUp, dotFwd);
  }

  const Cl = liftCoefficient(alpha);
  const Cd_induced = (Cl * Cl) / (Math.PI * e * AR);
  const Cd = Cd0 + Cd_induced;

  // Lift acts perpendicular to velocity in the plane of symmetry (local up
  // projected away from velocity direction).
  const liftMag = dynamicPressure * S * Cl;
  // Drag acts opposite to velocity.
  const dragMag = dynamicPressure * S * Cd;

  // Thrust along forward axis, scaled by density ratio
  const thrust = input.throttle * maxThrust * (rho / SEA_LEVEL_DENSITY);

  // ── Build force vector ────────────────────────────────────────────────

  const force = new THREE.Vector3();

  // Thrust — along forward
  force.addScaledVector(forward, thrust);

  if (V > 1) {
    const velDir = state.velocity.clone().normalize();

    // Lift — perpendicular to velocity, in the symmetry plane
    const liftDir = up.clone().addScaledVector(velDir, -up.dot(velDir));
    const liftDirLen = liftDir.length();
    if (liftDirLen > 0.001) {
      liftDir.divideScalar(liftDirLen);
      force.addScaledVector(liftDir, liftMag);
    }

    // Drag — opposite to velocity
    const dragDir = velDir.clone().negate();
    force.addScaledVector(dragDir, dragMag);

    // ── Sideslip lateral force (weathervane effect) ───────────────────
    //
    // When the aircraft heading differs from the velocity direction, the
    // fuselage and vertical tail present a lateral area to the airflow.
    // This generates a side force proportional to:
    //   - dynamic pressure (so it's weak at low speed / stall)
    //   - sideslip angle β (angle between velocity and nose in the yaw plane)
    //
    // The force pushes the velocity toward alignment with the heading.
    // It's NOT instant — it's an aerodynamic force that takes time, and
    // at low speed / stall it has almost no effect.

    const dotRight = velDir.dot(right);
    const beta = Math.asin(clamp(dotRight, -1, 1));

    // Side force: q × S × Cy_β × β, applied perpendicular to velocity
    // in the lateral direction, opposing the sideslip
    const sideForceMag = dynamicPressure * S * SIDE_FORCE_COEFF * beta;

    // Direction: component of right axis perpendicular to velocity
    const sideDir = right.clone().addScaledVector(velDir, -dotRight);
    const sideDirLen = sideDir.length();
    if (sideDirLen > 0.001) {
      sideDir.divideScalar(sideDirLen);
      // Negative sign: opposes sideslip (pushes velocity toward nose)
      force.addScaledVector(sideDir, -sideForceMag);
    }
  }

  // Weight — always down
  force.y -= mass * GRAVITY;

  // ── 3. Integration (semi-implicit Euler) ──────────────────────────────

  const accel = force.divideScalar(mass);
  state.velocity.addScaledVector(accel, safeDt);
  state.position.addScaledVector(state.velocity, safeDt);

  // Ground collision — simple clamp
  if (state.position.y < 0) {
    state.position.y = 0;
    // Kill downward velocity
    if (state.velocity.y < 0) {
      state.velocity.y = 0;
    }
  }

  // ── 4. Aerodynamic yaw (weathervane rotation) ──────────────────────
  //
  // The vertical tail creates a YAWING MOMENT that rotates the aircraft
  // nose toward the velocity direction.  This is what makes banked turns
  // actually change heading:
  //   bank → tilted lift curves velocity → sideslip develops →
  //   aero yaw rotates nose to follow → heading changes.
  //
  // Rate is proportional to β × (q / q_cruise), so it's strong at
  // cruise and negligible in a stall.

  const Vnew = state.velocity.length();
  if (Vnew > 1) {
    const velDirNew = state.velocity.clone().normalize();
    const dotRightNew = velDirNew.dot(right);
    const betaNew = Math.asin(clamp(dotRightNew, -1, 1));

    const qScale = dynamicPressure / Q_CRUISE;
    const aeroYawRate = betaNew * AERO_YAW_COEFF * qScale;

    // Apply yaw rotation around the aircraft's local up axis
    // Negative sign: rotates nose toward velocity (reduces β)
    const yawDq = new THREE.Quaternion();
    yawDq.setFromAxisAngle(up, -aeroYawRate * safeDt);
    quat.premultiply(yawDq);
    quat.normalize();
    state.rotation.setFromQuaternion(quat, "YXZ");

    // Recompute forward for heading after the yaw correction
    forward.set(0, 0, -1).applyQuaternion(quat);
  }

  // ── 5. Derived values ─────────────────────────────────────────────────

  state.airspeed = Vnew;
  state.altitude = state.position.y;

  // Heading: angle of the forward vector projected onto the XZ plane,
  // measured clockwise from north (+Z → 0, +X → 90).
  const fwd2d = new THREE.Vector2(forward.x, forward.z);
  if (fwd2d.length() > 0.001) {
    let hdg = Math.atan2(forward.x, forward.z) * (180 / Math.PI);
    if (hdg < 0) hdg += 360;
    state.heading = hdg;
  }

  state.throttle = input.throttle;

  return state;
}

// ── Factory ────────────────────────────────────────────────────────────────

/**
 * Create an initial aircraft state — airborne, heading north at cruise speed.
 */
export function createInitialState(): AircraftState {
  // Trim angle of attack: the nose-up pitch needed so Lift ≈ Weight at cruise.
  // At 60 m/s, 1000 m altitude:  α_trim ≈ 3° ≈ 0.053 rad.
  // Without this the aircraft starts at α=0 → Cl=0 → zero lift, and banking
  // has no effect because there is no lift vector to tilt.
  const TRIM_ALPHA = 0.053;

  // Euler "YXZ": x=pitch, y=yaw.  Positive x = nose up.
  // Yaw=PI faces the aircraft north (+Z).
  const rotation = new THREE.Euler(TRIM_ALPHA, Math.PI, 0, "YXZ");

  // Velocity is horizontal (north) at cruise speed.
  // The slight pitch-up gives α = TRIM_ALPHA between forward and velocity.
  const cruiseSpeed = 60;
  const velocity = new THREE.Vector3(0, 0, cruiseSpeed); // horizontal, north

  return {
    position: new THREE.Vector3(0, 1000, 0),
    rotation,
    velocity,
    airspeed: cruiseSpeed,
    altitude: 1000,
    heading: 0,
    throttle: 0.65,
    angularVelocity: new THREE.Vector3(),
  };
}

File: packages/wasm/src/physics/flight_model.rs
use bevy::prelude::*;
use std::f32::consts::PI;

use super::atmosphere::{self, G, RHO_SEA_LEVEL};
use crate::aircraft::Aircraft;

/// Maximum lift coefficient before stall.
const CL_MAX: f32 = 1.5;

/// Stall angle in radians (~15 degrees).
const STALL_ANGLE: f32 = 0.2618; // ~15 degrees

/// Compute coefficient of lift as a function of angle of attack.
///
/// Cl = 2*pi*alpha for small angles (thin airfoil theory).
/// After stall angle (~15 deg), lift drops off.
fn coefficient_of_lift(alpha: f32) -> f32 {
    let alpha_clamped = alpha.clamp(-0.5, 0.5);
    if alpha_clamped.abs() < STALL_ANGLE {
        // Linear region: Cl = 2*pi*alpha
        let cl = 2.0 * PI * alpha_clamped;
        cl.clamp(-CL_MAX, CL_MAX)
    } else {
        // Post-stall: lift drops off
        let sign = alpha_clamped.signum();
        let beyond_stall = (alpha_clamped.abs() - STALL_ANGLE) / (0.5 - STALL_ANGLE);
        let cl_at_stall = 2.0 * PI * STALL_ANGLE;
        let cl = cl_at_stall * (1.0 - 0.6 * beyond_stall);
        sign * cl.max(0.2)
    }
}

/// Compute the angle of attack from velocity and aircraft orientation.
///
/// Alpha is the angle between the velocity vector and the aircraft's
/// forward direction, measured in the aircraft's pitch plane.
fn compute_alpha(velocity_local: Vec3) -> f32 {
    let speed = velocity_local.length();
    if speed < 1.0 {
        return 0.0;
    }
    // In Bevy local space, forward is -Z, up is +Y.
    // For straight-and-level flight, velocity_local ≈ (0, 0, -speed).
    // alpha = atan2(-y, -z) so that forward flight gives alpha = 0.
    (-velocity_local.y).atan2(-velocity_local.z)
}

/// Main flight physics system.
///
/// Computes lift, drag, thrust, and weight forces, then integrates
/// velocity. All computation is in SI units.
pub fn update_flight_physics(
    time: Res<Time>,
    mut query: Query<(&mut Aircraft, &Transform)>,
) {
    let dt = time.delta_secs();
    if dt <= 0.0 || dt > 0.1 {
        return; // Skip bad timesteps
    }

    for (mut aircraft, transform) in query.iter_mut() {
        let altitude = transform.translation.y;
        let rho = atmosphere::density(altitude);

        // Aircraft local axes from its rotation
        let forward = transform.forward().as_vec3();
        let up = transform.up().as_vec3();
        let _right = transform.right().as_vec3();

        // Transform velocity into local space for aerodynamic calculations
        let rotation = transform.rotation;
        let velocity_local = rotation.inverse() * aircraft.velocity;
        let speed = aircraft.velocity.length();

        // Dynamic pressure: q = 0.5 * rho * V^2
        let q = 0.5 * rho * speed * speed;

        // Angle of attack
        let alpha = compute_alpha(velocity_local);

        // ---- LIFT ----
        // Lift = q * S * Cl(alpha)
        // Lift acts perpendicular to velocity, in the plane of aircraft up and velocity
        let cl = coefficient_of_lift(alpha);
        let lift_magnitude = q * aircraft.wing_area * cl;

        // Lift direction: perpendicular to velocity in the aircraft's pitch plane
        let lift_dir = if speed > 1.0 {
            let vel_normalized = aircraft.velocity.normalize();
            // Lift is perpendicular to velocity, biased toward aircraft up
            let lift_raw = up - vel_normalized * up.dot(vel_normalized);
            let len = lift_raw.length();
            if len > 0.001 {
                lift_raw / len
            } else {
                Vec3::Y
            }
        } else {
            Vec3::Y
        };

        let lift_force = lift_dir * lift_magnitude;

        // ---- DRAG ----
        // Cd = Cd0 + Cl^2 / (pi * e * AR)
        let induced_drag_coeff =
            cl * cl / (PI * aircraft.oswald_efficiency * aircraft.aspect_ratio);
        let cd = aircraft.cd0 + induced_drag_coeff;
        let drag_magnitude = q * aircraft.wing_area * cd;

        // Drag acts opposite to velocity
        let drag_force = if speed > 0.1 {
            -aircraft.velocity.normalize() * drag_magnitude
        } else {
            Vec3::ZERO
        };

        // ---- THRUST ----
        // Thrust = throttle * max_thrust * (rho / rho_sea_level)
        // Thrust acts along aircraft forward axis
        let density_ratio = rho / RHO_SEA_LEVEL;
        let thrust_magnitude = aircraft.throttle * aircraft.max_thrust * density_ratio;
        let thrust_force = forward * thrust_magnitude;

        // ---- WEIGHT ----
        let weight_force = Vec3::new(0.0, -aircraft.mass * G, 0.0);

        // ---- SUM FORCES ----
        let total_force = lift_force + drag_force + thrust_force + weight_force;

        // ---- ACCELERATION AND INTEGRATION ----
        let acceleration = total_force / aircraft.mass;
        aircraft.velocity += acceleration * dt;

        // Clamp velocity to prevent numerical instability
        let max_speed = 200.0; // m/s (~389 knots)
        if aircraft.velocity.length() > max_speed {
            aircraft.velocity = aircraft.velocity.normalize() * max_speed;
        }

        // Ground interaction: if on or below ground, prevent negative Y velocity
        // and provide a ground reaction force
        if altitude <= 0.5 {
            if aircraft.velocity.y < 0.0 {
                aircraft.velocity.y = 0.0;
            }
            // Add ground friction when on the ground
            if speed > 0.1 {
                let friction = -aircraft.velocity.normalize() * 0.02 * aircraft.mass * G;
                aircraft.velocity += friction * dt;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cl_at_zero_alpha() {
        let cl = coefficient_of_lift(0.0);
        assert!(cl.abs() < 0.001);
    }

    #[test]
    fn cl_linear_region() {
        let alpha = 0.1; // ~5.7 degrees
        let cl = coefficient_of_lift(alpha);
        let expected = 2.0 * PI * alpha;
        assert!((cl - expected).abs() < 0.01);
    }

    #[test]
    fn cl_post_stall_drops() {
        let cl_at_stall = coefficient_of_lift(STALL_ANGLE);
        let cl_past_stall = coefficient_of_lift(0.4);
        assert!(cl_past_stall < cl_at_stall);
    }

    #[test]
    fn alpha_straight_and_level() {
        // Velocity purely in forward direction (local -Z in Bevy)
        let alpha = compute_alpha(Vec3::new(0.0, 0.0, -60.0));
        assert!(alpha.abs() < 0.001);
    }

    #[test]
    fn alpha_nose_up() {
        // Velocity has downward component in local space = positive alpha (nose above flight path)
        let alpha = compute_alpha(Vec3::new(0.0, -10.0, -60.0));
        assert!(alpha > 0.0);
    }
}
